agent/leafcert/generate.go

```diff
@@ -0,0 +1,345 @@
+func (m *Manager) attemptLeafRefresh(
+	req *ConnectCALeafRequest,
+	existing *structs.IssuedCert,
+	state fetchState,
+) (*structs.IssuedCert, fetchState, error) {
+	if req.MaxQueryTime <= 0 {
+		req.MaxQueryTime = DefaultQueryTimeout
+	}
+
+	// Handle brand new request first as it's simplest.
+	if existing == nil {
+		return m.generateNewLeaf(req, state, true)
+	}
+
+	// We have a certificate in cache already. Check it's still valid.
+	now := time.Now()
+	minExpire, maxExpire := calculateSoftExpiry(now, existing)
+	expiresAt := minExpire.Add(lib.RandomStagger(maxExpire.Sub(minExpire)))
+
+	// Check if we have been force-expired by a root update that jittered beyond
+	// the timeout of the query it was running.
+	if !state.forceExpireAfter.IsZero() && state.forceExpireAfter.Before(expiresAt) {
+		expiresAt = state.forceExpireAfter
+	}
+
+	if expiresAt.Equal(now) || expiresAt.Before(now) {
+		// Already expired, just make a new one right away
+		return m.generateNewLeaf(req, state, false)
+	}
+
+	// If we called Get() with MustRevalidate then this call came from a non-blocking query.
+	// Any prior CA rotations should've already expired the cert.
+	// All we need to do is check whether the current CA is the one that signed the leaf. If not, generate a new leaf.
+	// This is not a perfect solution (as a CA rotation update can be missed) but it should take care of instances like
+	// see https://github.com/hashicorp/consul/issues/10871, https://github.com/hashicorp/consul/issues/9862
+	// This seems to me like a hack, so maybe we can revisit the caching/ fetching logic in this case
+	if req.MustRevalidate {
+		roots, err := m.rootsReader.Get()
+		if err != nil {
+			return nil, state, err
+		}
+		if activeRootHasKey(roots, state.authorityKeyID) {
+			return nil, state, nil
+		}
+
+		// if we reach here then the current leaf was not signed by the same CAs, just regen
+		return m.generateNewLeaf(req, state, false)
+	}
+
+	// We are about to block and wait for a change or timeout.
+
+	// Make a chan we can be notified of changes to CA roots on. It must be
+	// buffered so we don't miss broadcasts from rootsWatch. It is an edge trigger
+	// so a single buffer element is sufficient regardless of whether we consume
+	// the updates fast enough since as soon as we see an element in it, we will
+	// reload latest CA from cache.
+	rootUpdateCh := make(chan struct{}, 1)
+
+	// The roots may have changed in between blocking calls. We need to verify
+	// that the existing cert was signed by the current root. If it was we still
+	// want to do the whole jitter thing. We could code that again here but it's
+	// identical to the select case below so we just trigger our own update chan
+	// and let the logic below handle checking if the CA actually changed in the
+	// common case where it didn't it is a no-op anyway.
+	rootUpdateCh <- struct{}{}
+
+	// Subscribe our chan to get root update notification.
+	m.rootWatcher.Subscribe(rootUpdateCh)
```

Comment:
I wonder if this bug already exists in the code this is sourced from.

Reply:
Finally getting back to this, isn't
> Easy - have rootsWatcher.Subscribe immediately send a struct{} to the chan to fake an update. Then we let the logic in this funcs loop ignore because the active signing authority key has not changed.
already done by line 107?

Reply:
Yep, it is.